File: js/src/messageInterface.ts
import * as yup from 'yup';

export interface IMessage {
    type: string;
    data?: any;
    ID?: string;
    render_params?: any;
    methodName?: string;
    target?: string;
    args?: any[];
    kwargs?: any;
    component_index?: number;
    repr_index?: number;
    last_child?: boolean;
    reconstruc_color_scheme?: boolean;
    movie_making?: boolean;
    buffers?: any[];
    content?: any;
}

export const IMessageSchema = yup.object().shape({
    type: yup.string().required(),
    data: yup.mixed(),
    ID: yup.string(),
    render_params: yup.mixed(),
    methodName: yup.string(),
    target: yup.string(),
    args: yup.array(),
    kwargs: yup.object(),
    component_index: yup.number(),
    repr_index: yup.number(),
    last_child: yup.boolean(),
    reconstruc_color_scheme: yup.boolean(),
    movie_making: yup.boolean(),
    buffers: yup.array(),
    content: yup.mixed(),
});

File: nglview/message_interface.py
from pydantic import BaseModel
from typing import Any, List, Optional

class IMessage(BaseModel):
    type: str
    data: Optional[Any] = None
    ID: Optional[str] = None
    render_params: Optional[Any] = None
    methodName: Optional[str] = None
    target: Optional[str] = None
    args: Optional[List[Any]] = None
    kwargs: Optional[dict] = None
    component_index: Optional[int] = None
    repr_index: Optional[int] = None
    last_child: Optional[bool] = None
    reconstruc_color_scheme: Optional[bool] = None
    movie_making: Optional[bool] = None
    buffers: Optional[List[Any]] = None
    content: Optional[Any] = None